lib/Dialect/SV/SVOps.cpp

```diff
@@ -1055,6 +1057,130 @@ LogicalResult PAssignOp::verify() {
+              if (!constant)
+                return std::nullopt;
+              return ArraySlice{/*array=*/arrayIndex.getInput(),
+                                /*start=*/constant.getValue(),
+                                /*end=*/constant.getValue() + 1};
+            })
+        .Case<hw::ArraySliceOp>(
+            [](hw::ArraySliceOp slice) -> std::optional<ArraySlice> {
+              auto constant =
+                  slice.getLowIndex().getDefiningOp<hw::ConstantOp>();
+              if (!constant)
+                return std::nullopt;
+              return ArraySlice{
+                  /*array=*/slice.getInput(), /*start=*/constant.getValue(),
+                  /*end=*/
+                  constant.getValue() +
+                      hw::type_cast<hw::ArrayType>(slice.getType()).getSize()};
+            })
+        .Case<sv::IndexedPartSelectInOutOp>(
+            [](sv::IndexedPartSelectInOutOp index)
+                -> std::optional<ArraySlice> {
+              auto constant = index.getBase().getDefiningOp<hw::ConstantOp>();
+              if (!constant || index.getDecrement())
+                return std::nullopt;
+              return ArraySlice{/*array=*/index.getInput(),
+                                /*start=*/constant.getValue(),
+                                /*end=*/constant.getValue() + index.getWidth()};
+            })
+        .Default([](auto) { return std::nullopt; });
+  }
+};
+} // namespace
+
+static std::optional<std::pair<ArraySlice, ArraySlice>>
+getAssignedRange(Operation *op) {
+  auto srcRange = ArraySlice::getArraySlice(op->getOperand(1));
+  if (!srcRange)
+    return std::nullopt;
+  auto destRange = ArraySlice::getArraySlice(op->getOperand(0));
+  if (!destRange)
+    return std::nullopt;
+
+  return std::make_pair(*destRange, *srcRange);
+}
+
+// This canonicalization merges neiboring assignments of array elements into
+// array slice assignments. e.g.
+// a[0] <= b[1]
+// a[1] <= b[2]
+// ->
+// a[1:0] <= b[2:1]
+template <typename AssignTy>
+static LogicalResult mergeNeiboringAssignments(AssignTy op,
+                                               PatternRewriter &rewriter) {
+  // Check that a next operation is a same kind of the assignment.
+  AssignTy nextAssign = dyn_cast_or_null<AssignTy>(op->getNextNode());
+  if (!nextAssign)
+    return failure();
+
+  // Get assigned ranges of each assignment.
+  auto assignedRange = getAssignedRange(op);
+  auto nextAssignedRange = getAssignedRange(nextAssign);
+  if (!assignedRange || !nextAssignedRange)
+    return failure();
+
+  auto [dest, src] = *assignedRange;
+  auto [nextDest, nextSrc] = *nextAssignedRange;
+
+  // Check that these assignments are mergable.
+  if (!dest.isMergableSlice(nextDest) || !src.isMergableSlice(nextSrc))
+    return failure();
+
+  if (dest.start.getBitWidth() > 63)
+    return failure();
+  auto width = nextDest.end.getZExtValue() - dest.start.getZExtValue();
```

Comment:
`IndexedPartSelectInOutOp` seems to expect a `I32Attr` for the width. Can be do the subtraction on `APInt` and check if the result fits in 32 bits before passing it on to the op?

Reply:
I changed ArraySlice to have `width` explicitly so we can calculate the width by adding two widths. It should not overflow because `width1+width2< array.size`.